docs/proposals/scheduledjob.md

```diff
@@ -216,6 +208,144 @@ In the above example:
+   `"name":"nightly-earnings-report","uid":"5ef034e0-1890-11e6-8935-42010af0003e","apiVersion":...`
+   This serialization contains the UID of the parent.  This is used to match the Job to the SJ that created
+   it.
+
+## Updates to ScheduledJobs
+
+If the schedule were updated on a ScheduledJob it is hard for the controller to reason about
+whether a past job is intended to fulfill the current schedule or not.
+
+The controller matches Jobs to their ScheduledJob via the created-by annotation, which has a UID.
+If you delete a ScheduledJob and make an SJ with a the same name, the controller
+will not consider executions from the previous UID when making decisions about what executions to
+start, or status, etc.  When you replace, the old SJ is deleted, so you lose that history.
+So, if you want to preserve the old status, suspend the old one, and make one with a new name,
+or make a note of the old status.
+
+
+## Fault-Tolerance
+
+### Starting Jobs in the face of controller failures
+
+If the process with the scheduledJob controller in it fails,
+and takes a while to restart, the scheduledJob controller
+may miss the time window and it is too late to start a job.
+
+With a single scheduledJob controller process, we cannot give
+very strong assurances about not missing starting jobs.
+
+With a suggested HA configuration, there are multiple controller
+processes, and they use master election to determine which one
+is active at any time.
+
+If the Job's StartingDeadlineSeconds is long enough, and the
+lease for the master lock is short enough, and other controller
+processes are running, then a Job will be started.
+
+TODO: consider hard-coding the minimum StartingDeadlineSeconds
+at say 1 minute.  Then we can offer a clearer guarantee,
+assuming we know what the setting of the lock lease duration is.
+
+### Ensuring jobs are run at most once
+
+There are three problems here:
+
+- ensure at most one Job created per "start time" of a schedule.
+- ensure that at most one Pod is created per Job
+- ensure at most one container start occurs per Pod
+
+#### Ensuring one Job
+
+Multiple jobs might be created in the following sequence:
+
+1. scheduled job controller sends request to start Job J1 to fulfill start time T.
+1. the create request is accepted by the apiserver and enqueued but not yet written to etcd.
+1. scheduled job controller crashes
+1. new scheduled job controller starts, and lists the existing jobs, and does not see one created.
+1. it creates a new one.
+1. the first one eventually gets written to etcd.
+1. there are now two jobs for the same start time.
+
+We can solve this in several ways:
+
+1. with three-phase protocol, e.g.:
+  1. controller creates a "suspended" job.
+  1. controller writes writes an annotation in the SJ saying that it created a job for this time.
+  1. controller unsuspends that job.
+1. by picking a deterministic name, so that at most one object create can succeed.
```

Comment:
Yeah, but the scheduledjob.status.active will contain `ObjectReference` to an active job, if one isn't there it will repeat the process removing that reference, since it didn't find a job with that name. The only thing we'll need to add is to check for the existence of Active jobs from status in the scheduled controller. (And I hope we're both talking about the steps from my first comment.)